server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestPutMappingAction.java

```diff
@@ -48,15 +66,47 @@ public RestChannelConsumer prepareRequest(final RestRequest request, final NodeC
 
         Map<String, Object> sourceAsMap = XContentHelper.convertToMap(request.requiredContent(), false,
             request.getXContentType()).v2();
-        if (MapperService.isMappingSourceTyped(MapperService.SINGLE_MAPPING_NAME, sourceAsMap)) {
-            throw new IllegalArgumentException("Types cannot be provided in put mapping requests");
+        if (request.getRestApiVersion() == RestApiVersion.V_7) {
+            final boolean includeTypeName = request.paramAsBoolean(INCLUDE_TYPE_NAME_PARAMETER,
+                DEFAULT_INCLUDE_TYPE_NAME_POLICY);
+            if (request.hasParam(INCLUDE_TYPE_NAME_PARAMETER)) {
+                deprecationLogger.compatibleApiWarning("put_mapping_with_types", TYPES_DEPRECATION_MESSAGE);
+            }
+            final String type = request.param("type");
+            if (includeTypeName == false &&
+                (type != null || isMappingSourceTyped(MapperService.SINGLE_MAPPING_NAME, sourceAsMap))) {
+                throw new IllegalArgumentException("Types cannot be provided in put mapping requests, unless " +
+                    "the include_type_name parameter is set to true.");
+            }
+            // TODO what about the type in the source? it should be skipped?
+            //  (different than in RestCreateIndexAction which expects the _doc)
+            Map<String, Object> mappingSource = prepareV7Mappings(includeTypeName, sourceAsMap);
```

Comment:
my point here is that when a put mapping request on endpoint `PUT {index}/_mapping`
with body
```
PUT my_index 
{
  "mappings": {
    "MY_TYPE": {
      "properties": {
        "name": {
          "properties": {
            "first": {
              "type": "text"
            }
          }
        },
        "user_id": {
          "type": "keyword"
        }
      }
    }
  }
}
```

could have MY_TYPE which I simply omit. The `putMappingRequest.source` will be provided with a map which starts with `properties`